Host UI errors

vitek-karas opened the pull request:
Implements reporting errors with message box for Windows GUI apps in `apphost`.

In `apphost`, if there's no console detected, error writing will be redirected to a buffer. Upon exit, if the error buffer is not empty, it will be showed on screen as a message box.
The error writer is propagated from `apphost` to `hostfxr` and thus to `hostpolicy`.

This solves the problem that GUI apps which don't have console die silently without reporting any errors if there are issus during host execution.

vitek-karas committed:
Implements reporting errors with message box for Windows GUI apps in apphost.

In apphost, if there's no console detected, error writing will be redirected to a buffer. Upon exit, if the error buffer is not empty, it will be showed on screen as a message box.
The error writer is propagated from apphost to hostfxr and thus to hostpolicy.

This solves the problem that GUI apps which don't have console die silently without reporting any errors if there are issus during host execution.

vitek-karas committed:
Initial changes  to enable tests

vitek-karas committed:
Implement tests for UI error reports.

Added writing errors to debug output always.

vitek-karas commented:
Most of #4768 is fixed by this

sdmaclea reviewed (approved):
LGTM

sdmaclea commented:
@jkotas Can you put a s second set of eyes on thios before I merge it.

jkotas commented on src/corehost/corehost.cpp:
```diff
@@ -338,5 +353,39 @@ int main(const int argc, const pal::char_t* argv[])
+    // Detect if the process has console
+    // Note that this will fail even if a GUI process has STDOUT redirected. So this pretty effectively detects
+    // if the process is a GUI process without a console window.
+    HANDLE con_out = ::CreateFileW(L"CONOUT$", GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING, 0, 0);
```
Do we need to close this handle when we open it successfully?

vitek-karas replied:
Good point - we probably should. We're opening it for full sharing, so in theory it should not matter much, but it's definitely cleaner.

jkotas commented on src/corehost/corehost.cpp:
```diff
@@ -338,5 +353,39 @@ int main(const int argc, const pal::char_t* argv[])
-    return run(argc, argv);
+#if defined(_WIN32) && defined(FEATURE_APPHOST)
+    // Detect if the process has console
+    // Note that this will fail even if a GUI process has STDOUT redirected. So this pretty effectively detects
```
Have we lifted this logic from somewhere else? I would love to use the prior art here.

vitek-karas replied:
I came up with this - well - I read the docs for CreateFile and GetStdHandle and I tested them on both WinExe and Exe in various scenarios. While doing this I did some searching, but could not find anything existing. But I admit I didn't try super hard (I didn't expect this to be a tricky problem, and so far it doesn't seem to be).

sdmaclea replied:
@vitek-karas If you want me to do this, I'd be happy to...

vitek-karas replied:
@sdmaclea It would be great if you could also try to search for prior art here. I just did it again, found different answers but none even close to the quality we need (the most common answer is - try Console.ReadLine, if it throws, there's no console... not very good).
In general it seems that the ability to open the console "file" is a good indication, as it's intentionally "special" in Windows. Note that it's designed to "ignore" redirected input/output streams which is what we want here for consistency.
The other alternative I'm aware of is to open the .exe itself and look in the PE header to determine if it's marked GUI or not.

vitek-karas replied:
There's one scenarios which things might become "interesting" - that is a Windows Service. I guess I'll have to test that case, I don't know if services get console (even hidden) or not. We would want those to behave as console apps though (services can't show UI by default).

sdmaclea replied:
I also looked for prior art, but like Vitek said it is very immature.  @jkotas how important is using prior art.

I have no idea how to test a windows service.  Maybe this can sit for a while

I am wondering if the right approach might be to open the header and look for the gui bit.  The code must exist somewhere.  It seems like there would be prior art of sorts.

vitek-karas replied:
We have the code which sets the GUI bit (our SDK does that, I actually wrote that according to the PE spec, it's pretty simple). So if we feel that checking the GUI bit is a better way - we can definitely do that - with one condition - I don't know (yet) how to get to the memory address of a loaded module (we would need to find the .exe module, find the address it's mapped to and be allowed to read through some of its pages to get to the GUI bit).

jkotas replied:
I do not know myself what is better. Thanks for doing the due diligence. The approach looks fine to me given what we have found.

vitek-karas replied:
I'm trying the service … will report back

vitek-karas replied:
Windows Service seems to be rather tricky. It behaves as a GUI app (no console or stdout/stderr available), but it also doesn't like message boxes (in fact when I tried it, the service got into a state where it can't be removed by the OS - probably waits for a reboot as it's blocking some process somewhere).

So I'll probably look into the GUI bit reading.

jkotas replied:
> we would need to find the .exe module, find the address it's mapped to

`GetModule(NULL)` does exactly that.

> l that checking the GUI bit is a better

The interesting question is whether it is:
- better to be too aggressive, (display the message box even when it is necessary and a headless app can get stuck because of it)
- or whether it is better to be too conservative and miss displaying the dialog some of the time

I think it is better to be too conservative category, at least to start with. Message boxes on headless are super annoying to workaround. Missing the diagnostic is much easier to workaround.

If we wanted to check the GUI bit, we may want to do it in addition to what is in the change today. It would make us more conservative. (Just exploring the options, I am not saying that we should do it.)

vitek-karas replied:
Thanks @danmosemsft 
I don't know about .NET Framework, but I kind of doubt it. I'll see if I can maybe hack it somehow to make it fail and test this.

Current plan:
Switch to use the "find the GUI bit in the PE headers" instead of trying to open the console streams. It is less prone to issues caused by some weird setup of the app. Services being one such example.
I do agree that it's better to be conservative and only show the message box when it should work, which should be the case with the GUI bit.

weltkante replied:
Some .NET Framework startup issues of services certainly go into the Windows event log, e.g. unhandled managed exceptions during startup. It's the first place we look at when a service doesn't start up. If its supported to write Windows services in .NET Core this definitely needs to be covered, otherwise diagnosing startup issues is going to be insane for the poor developers diagnosing customer errors. After all if it doesn't start up we can't add custom logging so the runtime needs to have *something* in place to diagnose those startup problems, even if its just some flag we can set telling the runtime to log to a file instead of the console.

vitek-karas replied:
In 3.0 we've already added the support to redirect all tracing (which includes all errors as well and you can control verbosity) into a file through environment variable. Not sure if this is the best solution for windows services, but at least something.
See #4585 .

vitek-karas committed:
Change the algorithm to look for the GUI bit in PE header instead of trying to open console.

The previous version which used console stream as an indicator of a GUI app can lead to issues in cases where there's no console even for a "console" app - for example Windows Services are like this, they are compiled as console apps but run without attached console.

So to overcome this, this change modifies the algorithm to look for the GUI bit in the PE header directly. It will try to show a message box only for apps which are marked as GUI and thus should run in environment where showing GUI is OK.

jkotas commented on src/corehost/corehost.cpp:
```diff
@@ -333,6 +337,51 @@ void buffering_trace_writer(const pal::char_t* message)
+
+    HMODULE module = ::GetModuleHandleW(NULL);
+    MODULEINFO module_info;
+    if (!::GetModuleInformation(::GetCurrentProcess(), module, &module_info, sizeof(module_info)))
```
This is unnecessary. You  can just cast `HMODULE` to `void*` to get the address.

vitek-karas replied:
I could not find anywhere in the docs that it's guaranteed that HMODULE is the base address. But I guess windows will never break that anyway.

vitek-karas replied:
Anyway - I need the call to get the size to make sure we don't overrun any buffers. I know it should never happen (since otherwise OS would not load the module), but I'd rather keep this code itself safe.

jkotas replied:
You have to trust Windows that it validated the PE file headers while loading the image. Otherwise, you cannot really make this safe. E.g. the PE file can be mapped with holes - checking the length does not guarantee that accessing the memory will succeed.

Here is equivalent code from CoreCLR where we trust Windows that it validated the PE file: https://github.com/dotnet/coreclr/blob/030a3ea9b8dbeae89c90d34441d4d9a1cf4a7de6/src/inc/pedecoder.inl#L116

`GetModuleInformation` is diagnostic kind of API. It does not feel right calling it during startup.

vitek-karas replied:
OK - makes sense - removed the range checks and all the code around it.

jkotas commented on src/corehost/corehost.cpp:
```diff
@@ -333,6 +337,51 @@ void buffering_trace_writer(const pal::char_t* message)
+bool get_windows_graphical_user_interface_bit()
+{
+    // The first two bytes of a PE file are a constant signature.
+    const UINT16 PEFileSignature = 0x5A4D;
```
You can use constants for these from the standard Windows headers. This one is `IMAGE_DOS_SIGNATURE`

jkotas commented on src/corehost/corehost.cpp:
```diff
@@ -333,6 +337,51 @@ void buffering_trace_writer(const pal::char_t* message)
+    const UINT16 PEFileSignature = 0x5A4D;
+
+    // The offset of the PE header pointer in the DOS header.
+    const int PEHeaderPointerOffset = 0x3C;
```
`IMAGE_DOS_HEADER`

jkotas commented on src/corehost/corehost.cpp:
```diff
@@ -333,6 +337,51 @@ void buffering_trace_writer(const pal::char_t* message)
+    const int PEHeaderPointerOffset = 0x3C;
+
+    // The offset of the Subsystem field in the PE header.
+    const int SubsystemOffset = 0x5C;
```
This offset is different between 32-bit and 64-bit systems. `IMAGE_OPTIONAL_HEADER`

vitek-karas replied:
Actually the offset is the same (I tried with offsetof and so on). I will change the code here to use the structs from windows headers. This offset (actually the entire method) was copied from the SDK's code which marks the apphost as GUI app, so I will leave that code as is (it's C# and thus has no access to the structs, but the offsets should be correct).

vitek-karas committed:
Use constants and structs defined in the windows headers for the PE header parsing.

vitek-karas committed:
Remove range checks and trust the OS to only load valid images.

jaredpar commented:
Test comment, please ignore

Most of #3435
